content/2018-10-31-xsimd.rst

```diff
@@ -0,0 +1,272 @@
+
+That's our base line. If we recompile it with ``-DUSE_XSIMD``, we get an extra speedup (AVX instructions are available on my laptop, and enabled by ``-march=native``).
+
+.. code-block:: sh
+
+    CC=clang CXX=clang++ python -m pythran.run arc_distance.py -O3 -march=native -DUSE_XSIMD
+    python -m perf timeit -s 'N = 10000 ; import numpy as np ; np.random.seed(0); t0, p0, t1, p1 = np.random.randn(N), np.random.randn(N), np.random.randn(N), np.random.randn(N); from arc_distance import arc_distance' 'arc_distance(t0, p0, t1, p1)'
+    .....................
+    Mean +- std dev: 199 us +- 4 us
+
+That's roughly 7 times faster. And using Pythran 0.8.7, the last release with boost.simd support, we have
+
+.. code-block:: sh
+
+    CC=clang CXX=clang++ python -m pythran.run arc_distance.py -O3 -march=native -DUSE_BOOST_SIMD
+    python -m perf timeit -s 'N = 10000 ; import numpy as np ; np.random.seed(0); t0, p0, t1, p1 = np.random.randn(N), np.random.randn(N), np.random.randn(N), np.random.randn(N); from arc_distance import arc_distance' 'arc_distance(t0, p0, t1, p1)'
+    .....................
+    Mean +- std dev: 284 us +- 8 us
+
+This is slightly slower, but within the same magnitude order. Out of curiosity, I ran the same three experiments using Clang 6 as a backend compiler and I get the following timings:
+
+.. code::
+
+    clang + boost.simd: 220 us +- 8 us
+    clang + xsimd:      273 us +- 11 us
+    clang:             1.41 ms +- 0.04 ms
+
+Interestingly, **on that example**, Clang generates better code for the boost.simd version. So let's be wary of hasty conclusion and just state that with both engines, I can get efficient vectorization of my code.
+
+Complex Numbers
+===============
+
+Thanks to xsimd, Pythran is now able to *naively* support complex number
+vectorization. I state *naively* because we don't support changing internal
+representation from array-of-struct to struct-of-array, as we stick to numpy's
+layout. Still that's something new for Pythran as showcased by the following kernel:
+
+.. code-block:: python
+
+    #pythran export normalize_complex_arr(complex[])
+
+    import numpy as np
+    def normalize_complex_arr(a):
+        a_oo = a - a.real.min() - 1j*a.imag.min() # origin offsetted
+        return a_oo/np.abs(a_oo).max()
+
+
+Pythran provides a vectorized version of ``np.min`` and ``np.max`` operators, so thanks to complex support, it should provide some decent acceleration. Let's check that!
+
+.. code-block:: sh
+
+    CC=gcc CXX=g++ pythran -march=native -O3 normalize_complex_arr.py
+    python -m perf timeit -s 'import numpy as np; np.random.seed(0); N = 100000; x = np.random.random(N) + 1j *  np.random.random(N); from normalize_complex_arr import normalize_complex_arr' 'normalize_complex_arr(x)'
+    .....................
+    Mean +- std dev: 2.84 ms +- 0.01 ms
+
+
+And with vectorization on (note that the calls to ``np.min()`` do not involve complex numbers though.
```

Comment:
I find it a bit confusing because you want to show speed up thanks to complex vectorization but you mention a non complex function. I think stating that "with vectorization enabled, not only the real parts of the ``normalize_complex_arr`` function are accelerated, but the whole function is", could be clearer; there might be a better formulation than mine, but the idea is to point out which parts of the function are now accelerated thanks to complex vectorisation support.